### PR TITLE
Implement (try)FindLastSliceIndex

### LINKED ISSUE #574: [API Proposal] String.(try)FindLastSliceIndex
We have String.{try}FindSliceIndex as an alternative to C# String.IndexOf(string), I could use also the equivalent of String.LastIndexOf(string).

```fs
module String;

    let findLastSliceIndex (slice: string) (source: string) =
        let index = source.LastIndexOf slice
        if index = -1 then
            ArgumentException("The specified substring was not found in the string.") |> raise
        else
            index
            
    let tryFindLastSliceIndex (slice: string) (source: string) =
        let index = source.LastIndexOf slice
        if index = -1 then None else Some index
```

I can put up a PR if approved.

### PR DESCRIPTION
**Pull request by rzikm**
Closes #574

I was not able to build the test project, so I rely on CI to run them.

### CODE REVIEW

**Comment by wallymathieu** [comment]
@rzikm Hi! What kind of issues around running the tests do you have?

**Comment by gusty** [comment]
@rzikm That's fine, I also rely mostly in CIs run. You can keep adding commits, I will squash in the end.

Also, if you want to run them locally do it in Release mode, I don't think they will compile in DEBUG mode due to an old F# compiler bug.

**Comment by gusty** [comment]
> I also got snagged because there is more than one WrappedListD type and I added a function to only one of them. Maybe there is a room for internal cleanup?

AFAIK There's only one, which is in General.fs line 70

The main purpose of these types was to test overloads paths, but @wallymathieu recently added a better mechanism for it in #557 

Still these wrapper types are sometimes useful to perform some tests as they are user-defined classes.

Independently from that, cleanups are always welcome, but then you have to consider either doing them in a separate PR or in a separate commit and then squash all the other unrelated commits.

My advice is a separate commit so you can draft it first and we can give early feedback.

BTW: when doing a separate PR or opening a new issue please mention me because somehow I never managed to get notifications for them until I participate.

**Comment by rzikm** [comment]
> AFAIK There's only one, which is in General.fs line 70

The other one is in https://github.com/fsprojects/FSharpPlus/blob/54ffdcc9bf2638bca3d270e35c4e3de022ecc7d0/tests/FSharpPlusFable.Tests/FSharpTests/General/Util.fs#L67

**Comment by gusty** [comment]
Oh, but those are Fable tests.
Fable tests are independent, they have their own set of wrappers.

We can unify some stuff, but sometimes it gets complicated.

**Comment by gusty** [comment]
This seems to be ready to merge. The failure in appveyor is a transient one.

File: src/FSharpPlus/Internals.fs [hunk 1]
```diff
@@ -419,6 +419,97 @@ module FindSliceIndex =
...
+                    let last = if cache.SequenceEqual slice then index - slice.Length + 1 else last
+                    cache.Dequeue() |> ignore
+                    go last (index + 1)
+                else go last (index + 1)
+            else last
+        go -1 0
+    let sequenceEqual (a: _ seq) (b: _ seq) = a.SequenceEqual b
+    #else
+    let internal sequenceEqual (a: _ seq) (b: _ seq) :bool = Seq.compareWith Operators.compare a b = 0
+    module internal Q=
```
Comment (gusty) [inline src/FSharpPlus/Internals.fs:446]:
```suggestion
    module internal Q =
```

File: src/FSharpPlus/Internals.fs [hunk 2]
```diff
@@ -419,6 +419,97 @@ module FindSliceIndex =
...
+    type Queue<'T> () =
+        let mutable q : Q.queue<'T> = Q.empty
+        interface IEnumerable<'T> with
+            member _.GetEnumerator () = let s = Q.toSeq q in s.GetEnumerator()
+        interface IEnumerable with
+            member _.GetEnumerator () = let s = Q.toSeq q in s.GetEnumerator() :> IEnumerator
+        member _.Enqueue (v) = q <- Q.enqueue q v
+        member _.Dequeue () =
+            let (dequeued, next) = Q.dequeue q in q <- next
+            match dequeued with | Some v -> v | None -> failwith "Empty queue!"
```
Comment (gusty) [inline src/FSharpPlus/Internals.fs:478]:
```suggestion
            match dequeued with | Some v -> v | None -> invalidOp "Empty queue!"
```

**Review by gusty — CHANGES_REQUESTED** [review]
Looks good to me. Just 2 nits.
In fact, there's room for optimizations by using ListCollector and possibly ArrayCollector, but to be fair, the same would apply to `findSliceIndex`, so it's better to leave both for a separate PR.

Thanks for this contribution !

**Review by gusty — APPROVED** [review]

**Pull request merged**